File: Array/Algorithms/Java/sorting/quickSort.java
// Optimized Bubble sort in Java

import java.util.Arrays;

class Main {

  // perform the bubble sort
  private static void quickSort(int[] array, int start, int end) {
    if(start < end) {
        int pivot = partition(array, start, end);
        quickSort(array, start, pivot-1);
        quickSort(array, pivot+1, end);
    }
  }

  private static int partition(int[] array, int start, int end) {
    int pivot = end;
    int i = start - 1;
    for(int j = start; j <= end; j++) {
        if(array[j] <= array[pivot]) {
            i++;
            int temp = array[i];
            array[i] = array[j];
            array[j] = temp;
        }
    }
    return i;
  } 

  public static void main(String args[]) {
      
    int[] data = { 45, -3, 27, 0, 2, -8 };
    
    // call method using the class name
    Main.quickSort(data, 0, data.length-1);
    
    System.out.println("Sorted Array in Ascending Order:");
    System.out.println(Arrays.toString(data));
  }
}

File: Array/Algorithms/C++/sorting/quicksort.cpp
#include<iostream>
using namespace std;


int partition(int *a,int start,int end)
{
	int pivot,pindex,temp;
	pivot=a[end];
	pindex=start;
	for(int i=start;i<end;i++)
	{
		if(a[i]<=pivot)
		{
			temp=a[i];
			a[i]=a[pindex];
			a[pindex]=temp;
			pindex++;
		}
	}
	temp=a[end];
	a[end]=a[pindex];
	a[pindex]=temp;
	return pindex;
}


void quicksort(int *a,int start,int end)
{
	if(start<end)
	{
		int pindex=partition(a,start,end);
		quicksort(a,start,pindex-1);
		quicksort(a,pindex+1,end);
	}
}



int main()
{
	int n;
	cout<<"Enter number of elements";
	cin>>n;
	int a[n];
	cout<<"Enter elements";
	for(int i=0;i<n;i++)
	{
		cin>>a[i];
	}
	quicksort(a,0,n-1);
	for(int i=0;i<n;i++)
	{
		cout<<a[i];
	}
	return 0;
}